packages/ra-core/src/controller/record/WithRecord.tsx

```diff
@@ -14,14 +13,16 @@ import { useRecordContext } from './useRecordContext';
  *   </Show>
  * );
  */
-export const WithRecord = <RecordType extends RaRecord>({
+export const WithRecord = <RecordType extends Record<string, unknown> = any>({
     render,
 }: WithRecordProps<RecordType>) => {
     const record = useRecordContext<RecordType>();
     return record ? render(record) : null;
 };
 
-export interface WithRecordProps<RecordType extends RaRecord> {
-    render: (record: RecordType) => ReactElement;
+export interface WithRecordProps<
+    RecordType extends Record<string, unknown> = any
+> {
```

Comment:
Updated these to match #8964

Reply:
@fzaninotto @slax57 it might be helpful to create a `StrictRaRecord` type that we can share/extend from?

e.g.

```tsx
// ra-core/src/types.ts

export interface StrictRaRecord<IdentifierType extends Identifier = Identifier> {
    id: IdentifierType;
}

export interface RaRecord<IdentifierType extends Identifier = Identifier> extends StrictRaRecord {
    [key: string]: any;
}

// in other files…

export interface WithRecordProps<RecordType extends StrictRaRecord = RaRecord> { … }
export interface FunctionFieldProps<RecordType extends StrictRaRecord = RaRecord> { … }

Reply:
I'm not sure of the added value (extending a StrictRaRecord vs adding an Identifier by hand is about the same). Besides, it's a more general reflexion about struct null checks that we're working on in the long term.